src/geom/elem.C

```diff
@@ -356,6 +356,110 @@ const Elem * Elem::reference_elem () const
 
 
 Point Elem::centroid() const
+{
+  libmesh_do_once(libMesh::err
+                  << "Elem::centroid() has been deprecated. Replace with either "
+                  << "Elem::vertex_average() to maintain existing behavior, or "
+                  << "the more expensive Elem::true_centroid() "
+                  << "in cases where the true 'geometric' centroid is required."
+                  << std::endl);
+  libmesh_deprecated();
+
+  return Elem::vertex_average();
+}
+
+Point Elem::true_centroid() const
+{
+  // The base class implementation builds a finite element of the correct
+  // order and computes the centroid, c=(cx, cy, cz), where:
+  //
+  // [cx]            [\int x dV]
+  // [cy] := (1/V) * [\int y dV]
+  // [cz]            [\int z dV]
+  //
+  // using quadrature. Note that we can expand "x" in the FE space as:
+  //
+  // x = \sum_i x_i \phi_i
+  //
+  // where x_i are the nodal positions of the element and \phi_i are the
+  // associated Lagrange shape functions. This allows us to write the
+  // integrals above as e.g.:
+  //
+  // \int x dV = \sum_i x_i \int \phi_i dV
+  //
+  // Defining:
+  //
+  // V_i := \int \phi_i dV
+  //
+  // we then have:
+  //
+  // [cx]           [\sum_i x_i V_i]
+  // [cy] = (1/V) * [\sum_i y_i V_i]
+  // [cz]           [\sum_i z_i V_i]
+  //
+  // where:
+  // V = \sum_i V_i
+  //
+  // Derived element types can overload this method to compute
+  // the centroid more efficiently when possible.
+
+  // If this Elem has an elevated p_level, then we need to generate a
+  // barebones copy of it with zero p_level and call true_centroid()
+  // on that instead.  This workaround allows us to avoid issues with
+  // calling FE::reinit() with a default_order() FEType, and then
+  // having that order incorrectly boosted by p_level.
+  if (_p_level)
+    {
+      auto elem_copy = Elem::build(this->type());
```

Comment:
I originally didn't like this idea because of the expense of having to make a heap allocation inside a centroid calculation; it didn't occur to me that you *only* have to make the allocation in the uncommon case.  We still ought to fix this in FE at some point.

Reply:
> We still ought to fix this in FE at some point.

I didn't look into it too carefully, but it just seemed like it would be awkward. Adding p_level support in the first place was already a tricky dance of trying to find all the places where it needed to be accounted for, now we would need to go in and add a "oops, just kidding don't actually use p_level" flag to all of those places.